src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -219,73 +249,84 @@ static void CheckForExperimentStatus(ITextView textView, Document document)
-                        completionList.SuggestionModeItem.DisplayText,
-                        completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
-                            ? description
-                            : string.Empty)
-                    : null;
+                items = itemsBuilder.ToImmutableAndFree();
 
-            // Have to store the snapshot to reuse it in some projections related scenarios
-            // where data and session in further calls are able to provide other snapshots.
-            session.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                suggestionItemOptions = completionList.SuggestionModeItem != null
+                        ? new AsyncCompletionData.SuggestionItemOptions(
+                            completionList.SuggestionModeItem.DisplayText,
+                            completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                                ? description
+                                : string.Empty)
+                        : null;
 
-            // Store around the span this completion list applies to.  We'll use this later
-            // to pass this value in when we're committing a completion list item.
-            session.Properties.AddProperty(CompletionListSpan, completionList.Span);
+                if (!isExpanded)
+                {
+                    // Store around the span this completion list applies to.  We'll use this later
+                    // to pass this value in when we're committing a completion list item.
+                    session.Properties[CompletionListSpan] = completionList.Span;
 
-            // This is a code supporting original completion scenarios: 
-            // Controller.Session_ComputeModel: if completionList.SuggestionModeItem != null, then suggestionMode = true
-            // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
-            session.Properties.AddProperty(HasSuggestionItemOptions, suggestionItemOptions != null);
+                    // This is a code supporting original completion scenarios: 
+                    // Controller.Session_ComputeModel: if completionList.SuggestionModeItem != null, then suggestionMode = true
+                    // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
+                    session.Properties[HasSuggestionItemOptions] = suggestionItemOptions != null;
 
-            var excludedCommitCharacters = GetExcludedCommitCharacters(completionList.Items);
-            if (excludedCommitCharacters.Length > 0)
-            {
-                session.Properties.AddProperty(ExcludedCommitCharacters, excludedCommitCharacters);
+                    var excludedCommitCharacters = GetExcludedCommitCharacters(completionList.Items);
```

Comment:
do we not care about excluded commit characters when we send expanded item set?

Reply:
@ivanbasov Should I merge the chars from completionList.Items with the value of current `ExcludedCommitCharacters` property here?

Reply:
Yes, exactly. It seems to be the right idea. We exclude characters that are filter characters at least for 1 completion item. If we were called twice (for two different languages), we should merge their `ExcludedCommitCharacters`. Thank you, @genlu !

Reply:
Fixed. Now we merge those characters